### PR TITLE
Add rounding functions and refactor

### PR DESCRIPTION
**Pull request by falbrechtskirchinger**
This is another PR split out of #1594 and is **not intended to be merged as-is** (see explanation below).

Changes:
* Add a floating-point to integer rounding function `iround<IntType>()`.
* Replace uses of `std::lround()` with the cheaper `iround()`.
* Add a function for rounding integers to multiples of `N`.
* Modify the `Point` conversion constructor to round when converting floating-point element types to integers.

Additionally, the following change was made to discover the affected code paths by rounding in the `Point` conversion constructor:

The floating-point to integer conversion operator was replaced with two variants – one using rounding, one using plain casting – disambiguated via tag dispatch.

6 uses (excluding test code) were identified this way, and only 2 of those don't *require* rounding because the `Point<float>` was created from integer values.

if desired, a tag dispatch variant to disable rounding could be kept, but the benefits seem negligible.
Unless someone identifies a use case where rounding *has* to be disabled, I recommend deleting both tag dispatch variants and rounding all floating-point to integer conversions indiscriminately.

I did consider adding a new header, but `mathFuncs.h` already contains related functions (e.g., `roundedDiv()`).

To-do:
- [ ] Remove CI commit from this PR once merged. (#1596)
- [ ] Add unit tests.
- [ ] Decide whether to keep at least one tag dispatch overload around or delete both.

### CODE REVIEW

**Review by Flamefire — CHANGES_REQUESTED** [review]
Thanks I think it makes sense to round instead of truncate although it can be a pitfall. What exactly is this **required** for though?

Also have you double-checked all files where you added `mathFuncs.h` and removed `std::lround` if the `cmath` include can be removed?

File: libs/common/src/helpers/mathFuncs.cpp [hunk 1]
```diff
@@ -33,4 +33,11 @@ unsigned roundedDiv(unsigned dividend, unsigned divisor) noexcept
     // Which is the same as: floor((divident + 0.5 * divisor) / divisor) == floor((divident + divisor / 2) / divisor)
     return (dividend + (divisor / 2)) / divisor;
 }
+
+namespace detail {
+    void assertNotNegative(bool negative) noexcept
+    {
+        RTTR_Assert_Msg(!negative, "Floating-point value must not be negative when casting to unsigned integer type.");
```
[the remark this answers is not in the document]
Comment (falbrechtskirchinger) [reply libs/common/src/helpers/mathFuncs.cpp:40]: I wanted to avoid pulling in the rather heavy `RTTR_Assert.h` into this otherwise lightweight header.

File: libs/common/include/helpers/mathFuncs.h [hunk 2]
```diff
@@ -79,4 +80,27 @@ constexpr T inverseLerp(const T startVal, const T endVal, const T value) noexcep
 {
     return (value - startVal) / (endVal - startVal);
 }
+
+namespace detail {
+    void assertNotNegative(bool negative) noexcept;
+} // namespace detail
+
+/// Arithmetically round floating point values to integers
+/// Can produce slightly incorrect results for certain edge cases
```
[the remark this answers is not in the document]
Comment (falbrechtskirchinger) [reply libs/common/include/helpers/mathFuncs.h:89]: I should probably update that comment. There are cases where the rounding is more than *slightly* of. It depends on whether the number is representable by the floating-point type. Literals may be rounded incorrectly if they aren't representable exactly, for example. The integer type can overflow.
Most of these issues apply to `std::round()` as well (if the end result is narrowed by `static_cast` without checking).

There're supposedly more nebulous issues for numbers right next to `±x.5f` in the set of representable floating-point values that I didn't investigate any further. But assuming we're not also running into one of the issues above, these are *off-by-one* errors at most.

In the end, none of these are deal breakers in a video game, IMHO.

File: libs/common/include/helpers/mathFuncs.h [hunk 3]
```diff
@@ -79,4 +80,27 @@ constexpr T inverseLerp(const T startVal, const T endVal, const T value) noexcep
...
+    if(std::is_unsigned<IntType>::value)
+    {
+#ifndef NDEBUG
+        detail::assertNotNegative(val < 0);
+#endif
+        return static_cast<IntType>(val + FloatType(0.5)); // NOLINT(bugprone-incorrect-roundings)
+    } else
+    {
+        const auto sign = static_cast<IntType>((val > 0) - (val < 0));
+        return static_cast<IntType>(std::fabs(val) + FloatType(0.5)) * sign; // NOLINT(bugprone-incorrect-roundings)
```
[the remark this answers is not in the document]
Comment (falbrechtskirchinger) [reply libs/common/include/helpers/mathFuncs.h:103]: > I'd rather keep using the ` static_cast<T>(std::lround(...))` you replaced by this or is there a reason this is required for?

I've looked at the implementation of `lround` in glibc and thought it was ridiculous. It exploits knowledge of the floating-point memory layout and uses integer math + branching. Integer math performance on modern CPUs isn't stellar compared to single-precision floating-point.

> Edit: You wrote this is supposed to be faster? Is it indeed that much faster it is worth rolling a custom implementation? Especially as it requires suppressing a LINT warning which [states](https://clang.llvm.org/extra/clang-tidy/checks/bugprone/incorrect-roundings.html):
> 
> > It is unnecessarily slow.

Note the use of double precision. Another issue I've identified with this project and intend to clean up at some point. Don't use `double` unless you need extra precision. I think even the `std::chrono` stuff doesn't need it, but I have to check.

Is it ultimately worth it? I don't know. I'd be content if `Point` would round, removing the need for `+ PointF::all(.5f)` in #1594.

:man_shrugging:

File: libs/common/include/Point.h [hunk 4]
```diff
@@ -21,12 +22,28 @@ struct Point //-V690
     using ElementType = T;
     static_assert(std::is_arithmetic<ElementType>::value, "Requires an arithmetic type");
 
+    struct NoRounding_t
```
[the remark this answers is not in the document]
Comment (falbrechtskirchinger) [reply libs/common/include/Point.h:25]: Agreed. I like `Truncate`.

**Comment by falbrechtskirchinger** [comment]
> Thanks I think it makes sense to round instead of truncate although it can be a pitfall. What exactly is this **required** for though?

None of the calculations are so precise that I'd say it's required, but it does feel wrong to me to just truncate in all paces. I guess that's the reason why I've focused on making it fast at least.

> Also have you double-checked all files where you added `mathFuncs.h` and removed `std::lround` if the `cmath` include can be removed?

I did check once but would want to double-check if `iround()` is here to stay.

<hr>

I've already removed one function from this PR and I don't feel strongly about keeping `iround()` TBH. But `Point` should round by default for correctness' sake.

File: libs/common/src/helpers/mathFuncs.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Flamefire) [reply libs/common/src/helpers/mathFuncs.cpp:40]: It is not that heavy and already included in pretty much every second CU, so including it is almost free.

File: libs/common/include/helpers/mathFuncs.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (Flamefire) [reply libs/common/include/helpers/mathFuncs.h:103]: > I've looked at the implementation of `lround` in glibc and thought it was ridiculous...

As usual: It depends ;-) Unless we see that this rounding is a bottleneck IMO the clarity is worth more than some potentially (but not proven) minor speed advantage on a rare operation. Especially for maintainability of an FOSS project clarity is more important. 

> > > It is unnecessarily slow.
> 
> Note the use of double precision.

You mean in the example there? Yeah could be that the "slow" part doesn't apply to single precision as much. Still: Don't do more coding work which makes understanding the code harder unless it is required.

> Another issue I've identified with this project and intend to clean up at some point. Don't use `double` unless you need extra precision. I think even the `std::chrono` stuff doesn't need it, but I have to check.

True. But using `double` relieves the work of thinking **iff** you need the extra precision. So it is a trade-off avoiding extra design work given that modern CPUs have enough power such that single vs double precision often doesn't really matter (often the same unit is used and as we do only very few floating point ops I don't think we saturate them, so the extra precision is almost free)

> Is it ultimately worth it? I don't know. I'd be content if `Point` would round, removing the need for `+ PointF::all(.5f)` in #1594.

Ok. As I tend to the safe version in case of doubt (see motivation above) I'd say use `std::lround`. Having it in one place allows changing it later if we deem it worth it. And maybe a few tests for `iround` would be good, especially if you can imagine a few edge cases or so worth highlighting.

**Comment by Flamefire** [comment]
> but it does feel wrong to me to just truncate in all paces.

Initial motivation: Conversion between different Point types is explicit, so `Point<int>(floatPoint)` acts like `static_cast<int>(floatValue)` especially as that is what happens between integer Point-types

I guess the idea was that usually a position or size is used. And you don't want that to accidentally exceed a given size due to floating point (in)accuracy, so in those cases 1 px less might be better than 1px more.

> I don't feel strongly about keeping iround() TBH.

I think it is good to encapsulate the `static_cast(round(value))` thing into a function where we could place asserts. So IMO it is worth keeping.

> But Point should round by default for correctness' sake.

See the initial motivation: To me it isn't that clear if it by default should round or act like a cast. But having a ctor that rounds is useful as you identified. I'm fine with that being the default and I like the tag dispatch.   
For completeness: An alternative would be to require a rounding mode (e.g. enum: truncate or round) for float->int conversion of points.

Unless @Flow86 feels strongly either way I'd be fine with the currently discussed changes (i.e. use of `Truncate` and `std::lround`, round by default)

**Comment by falbrechtskirchinger** [comment]
Let me know if the additional error handling is too much. I chased a silly bug and that's what I ended up with before I noticed a simple typo.

File: libs/common/include/helpers/mathFuncs.h [hunk 5]
```diff
@@ -79,4 +81,19 @@ constexpr T inverseLerp(const T startVal, const T endVal, const T value) noexcep
...
 }
+
+// TODO can be constexpr in C++20
+/// Arithmetically round floating point values to integers
+template<typename IntType, typename FloatType,
+         std::enable_if_t<std::is_integral<IntType>::value && std::is_floating_point<FloatType>::value, int> = 0>
+inline IntType iround(const FloatType val) noexcept
+{
+    RTTR_Assert(!std::isnan(val));
+    RTTR_Assert(!std::isinf(val));
```
Comment (Flamefire) [inline libs/common/include/helpers/mathFuncs.h:92]: Good idea but I guess a combined check is enough:
```suggestion
    RTTR_Assert(!std::isfinite(val));
```

File: libs/common/include/helpers/mathFuncs.h [hunk 6]
```diff
@@ -79,4 +81,18 @@ constexpr T inverseLerp(const T startVal, const T endVal, const T value) noexcep
 {
     return (value - startVal) / (endVal - startVal);
 }
+
+// TODO can be constexpr in C++20
+/// Arithmetically round floating point values to integers
+template<typename IntType, typename FloatType,
+         std::enable_if_t<std::is_integral<IntType>::value && std::is_floating_point<FloatType>::value, int> = 0>
+inline IntType iround(const FloatType val) noexcept
```
Comment (Flamefire) [inline libs/common/include/helpers/mathFuncs.h:89]: Sorry missed that: The `inline` is redundant for template & constexpr functions. IIRC we follow this consistently so I don't want to introduce a different style.
```suggestion
IntType iround(const FloatType val) noexcept
```

**Review by Flamefire — APPROVED** [review]

**Comment by falbrechtskirchinger** [comment]
You know … I'm starting to think I wasn't imagining a division-by-zero issue in the FrameCounter …

`std::lround()` just happens to return `0` in that case in all relevant implementations.

**Comment by Flamefire** [comment]
Uff, sorry I thought this was done now and wanted to spare you the work to apply and rebase such that Github allows me to merge this. Feel free to force-push your version (with the inline-removal applied)

**Comment by falbrechtskirchinger** [comment]
Don't worry, it's fine. Just running tests locally to double-check, then I'll push.

**Review by Flamefire — APPROVED** [review]
